### ISSUE TITLE
Demo for Distributed Key Generation (DKG)

### ISSUE THREAD

**Issue by mpguerra** [ZcashFoundation/frost#252]
# Motivation

We want to simulate a execution of FROST using DKG to show potential FROST implementers how to use the DKG.

## Scope

Similar to ZcashFoundation/frost#238, but using the DKG instead.

The DKG demo must be run for each signer, in parallel (e.g. one execution of the same binary in different terminals).

At the end of the execution, the program must print all the information required for signing and verifying (using ZcashFoundation/frost#239 and ZcashFoundation/frost-zcash-demo#48)

## Specification

On startup, it will prompt for:

- The number of shares to generate (`MAX_PARTICIPANTS`)
- The threshold number of participants for the secret sharing scheme (`MIN_PARTICIPANTS`)
- The identifier of the current participant (an integer)

It will the run `part1` of the DKG. Then:

- Print the Round 1 `Package` (suggestion: in a single-line JSON to make it easy to copy&paste)
- Store the Round 2 `SecretPackage` in memory

The program keeps running, then prompts for:

- For every other participant:
  - The `SecretPackage` received from that participant. (Note that it could be easier to include the identifier inside the package itself. That would be a bad example, though, since the identifier must be deduced from the authenticated communication channel, which does not exist in the demo. We need to make it clearer to library users that the identifier should not be simply sent inside the channel since it could be spoofed.)

It will run `part2` of the DKG. Then:

- Print the Round 2 `Package` for each other participant
- Store the Round 2 `SecretPackage` in memory

The program keeps running, then prompts for:

- For every other participant:
  -  The `SecretPackage` received from that participant.

Finally, the program runs `part3` of the DKG using the `Package`s received in both round 1 and round 2 and the round 2 `SecretPackage. It then prints:

- The participant secret share
- The participant verifying share
- The group public key

**Comment by ThantZin-develope** [comment]
Prompts for the SecretPackage received from that participant? is not it Package broadcast from other participants? and that Package struct is not Printable?

**Comment by mpguerra** [comment]
This is just meant as a toy/demo implementation as an example and is not intended to be used in a production setting

**Comment by mpguerra** [comment]
Do we want to close this with frost-zcash-demo#36 or split it into implementation and test issues and just close the implementation part? I'm good with either option

**Comment by conradoplg** [comment]
> Do we want to close this with [ZcashFoundation/frost-zcash-demo#36](https://github.com/ZcashFoundation/frost-zcash-demo/pull/36) or split it into implementation and test issues and just close the implementation part? I'm good with either option

I don't mind either way, I think it's simpler to keep this open, I will probably work on the tests soonish
